pkg/controllers/nodepool/status/controller.go

```diff
@@ -37,61 +35,58 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
-	"sigs.k8s.io/karpenter/pkg/utils/resources"
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+	"sigs.k8s.io/karpenter/pkg/controllers/state"
+	operatorcontroller "sigs.k8s.io/karpenter/pkg/operator/controller"
 )
 
 var _ operatorcontroller.TypedController[*v1beta1.NodePool] = (*Controller)(nil)
 
+type nodePoolStatusReconciler interface {
+	Reconcile(context.Context, *v1beta1.NodePool) (reconcile.Result, error)
+}
+
 // Controller for the resource
 type Controller struct {
-	kubeClient client.Client
-	cluster    *state.Cluster
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+	counter       *Counter
+	readiness     *Readiness //TODO : Remove this when we have sub status conditions
 }
 
 // NewController is a constructor
-func NewController(kubeClient client.Client, cluster *state.Cluster) operatorcontroller.Controller {
+func NewController(kubeClient client.Client, cluster *state.Cluster, cloudProvider cloudprovider.CloudProvider) operatorcontroller.Controller {
 	return operatorcontroller.Typed[*v1beta1.NodePool](kubeClient, &Controller{
-		kubeClient: kubeClient,
-		cluster:    cluster,
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+		counter:       &Counter{cluster},
+		readiness:     &Readiness{cloudProvider: cloudProvider, kubeClient: kubeClient},
 	})
 }
 
 // Reconcile a control loop for the resource
 func (c *Controller) Reconcile(ctx context.Context, nodePool *v1beta1.NodePool) (reconcile.Result, error) {
-	// We need to ensure that our internal cluster state mechanism is synced before we proceed
-	// Otherwise, we have the potential to patch over the status with a lower value for the nodepool resource
-	// counts on startup
-	if !c.cluster.Synced(ctx) {
-		return reconcile.Result{RequeueAfter: time.Second}, nil
-	}
 	stored := nodePool.DeepCopy()
-	// Determine resource usage and update nodepool.status.resources
-	nodePool.Status.Resources = c.resourceCountsFor(v1beta1.NodePoolLabelKey, nodePool.Name)
+	var results []reconcile.Result
+	var errs error
+	for _, reconciler := range []nodePoolStatusReconciler{
```

Comment:
Do we need to couple these controllers or can they be separate controllers entirely?

Reply:
I think it makes sense to have these controllers coupled because both of them make update to nodepool status. We did something similar for the AWS provider where we coupled the nodeClass controllers that update status. Thoughts?